ocrd_cis/ocropy/clip.py

```diff
@@ -76,13 +74,10 @@ def process(self):
         # too. However, region-level clipping _must_ be run before region-level
         # deskewing, because that would make segments incomensurable with their
         # neighbours.
-        LOG = getLogger('processor.OcropyClip')
         level = self.parameter['level-of-operation']
-        assert_file_grp_cardinality(self.input_file_grp, 1)
-        assert_file_grp_cardinality(self.output_file_grp, 1)
 
         for (n, input_file) in enumerate(self.input_files):
```

Comment:
Shouldn't we migrate to `process_page_pcgts` here, too?

Reply:
Yes, everywhere. For now only binarize is migrated. I am just lagging with the migration since I have no working tests locally. The server from where the resources are downloaded is not available anymore. Hence, I try to adapt only things I understand and I am sure are the right things to do.

Reply:
Oops! Strange that Github pull the older release of our GT. Anyway, with https://github.com/OCR-D/gt_structure_text/pull/2 out of the way it should now suffice to change the base URL in 

https://github.com/MehmedGIT/ocrd_cis/blob/156d79fc051abeecf001cd6973e71c18efc659dd/tests/test_lib.bash#L8

to `https://github.com/OCR-D/gt_structure_text/releases/tag/v1.5.0/`

Reply:
And again, thanks for being so thorough!

Reply:
That still did not help.
```bash
(venv38-core) mm@MM-Notebook:~/repos/ocrd_cis$ make test
bash tests/run_add_zip_test.bash > /dev/null 2>&1
make: *** [Makefile:25: tests/run_add_zip_test.bash] Error 1
```

To get more detailed errors, I did:
```bash
(venv38-core) mm@MM-Notebook:~/repos/ocrd_cis$ bash tests/run_add_zip_test.bash
--2024-08-14 14:47:30--  https://github.com/OCR-D/gt_structure_text/releases/tag/v1.5.0//blumenbach_anatomie_1805.ocrd.zip
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘/home/mm/repos/ocrd_cis/download/blumenbach_anatomie_1805.ocrd.zip’

blumenbach_anatomie_1805.ocrd.zip                   [ <=>                                                                                                   ] 172,14K  --.-KB/s    in 0,08s   

2024-08-14 14:47:30 (2,16 MB/s) - ‘/home/mm/repos/ocrd_cis/download/blumenbach_anatomie_1805.ocrd.zip’ saved [176273]

14:47:31.446 INFO ocrd.workspace_bagger - Spilling /home/mm/repos/ocrd_cis/download/blumenbach_anatomie_1805.ocrd.zip to /tmp/tmp.QcrQegkntf/blumenbach_anatomie_1805
Traceback (most recent call last):
  File "/home/mm/venv38-core/bin/ocrd", line 8, in <module>
...
  File "/home/mm/repos/core/build/__editable__.ocrd-2.67.2-py3-none-any/ocrd_utils/os.py", line 74, in unzip_file_to_dir
    z = ZipFile(path_to_zip, 'r')
  File "/usr/lib/python3.8/zipfile.py", line 1271, in __init__
    self._RealGetContents()
  File "/usr/lib/python3.8/zipfile.py", line 1338, in _RealGetContents
    raise BadZipFile("File is not a zip file")
zipfile.BadZipFile: File is not a zip file
```
Seems the download fails for some reason since the size of the zip is much smaller. I have manually downloaded the zip and placed it where it is expected. Hooray, the tests started passing, but then:

```bash
(venv38-core) mm@MM-Notebook:~/repos/ocrd_cis$ make test
bash tests/run_add_zip_test.bash > /dev/null 2>&1
bash tests/run_alignment_test.bash > /dev/null 2>&1
make: *** [Makefile:25: tests/run_alignment_test.bash] Error 1
```

Reply:
I apologise – it should have read `https://github.com/OCR-D/gt_structure_text/releases/download/v1.5.0/` (i.e. download, not tag)

Reply:
That fixed it, yes. 

There are other small errors but I guess those are related now to the missing parts expected from core v3.0.0. E.g.:

```python
  File "/home/mm/venv38-core/lib/python3.8/site-packages/ocrd_cis/ocropy/segment.py", line 14, in <module>
    from shapely import set_precision
ImportError: cannot import name 'set_precision' from 'shapely' (/home/mm/venv38-core/lib/python3.8/site-packages/shapely/__init__.py)
```
However, that error does not occur in a different Python environment.

```python
  File "/home/mm/repos/core/build/__editable__.ocrd-2.67.2-py3-none-any/ocrd_utils/os.py", line 93, in get_ocrd_tool_json
    if 'resource_locations' not in ocrd_tool:
TypeError: argument of type 'NoneType' is not iterable
```

And this is just the missing `ocrd_tool` that should come from the `self._metadata` from core v3.0.0

Reply:
> `ImportError: cannot import name 'set_precision' from 'shapely' (/home/mm/venv38-core/lib/python3.8/site-packages/shapely/__init__.py)`

Can you check `pip show shapely`? Looks to me like it is still `<2`.

> "/home/mm/repos/core/build/__editable__.ocrd-2.67.2-py3-none-any/ocrd_utils/os.py"

Can you `git pull` the latest changes from #1240? In the most recent state, this file should be `/home/mm/repos/core/build/__editable__.ocrd-3.0.0a1-py3-none-any/ocrd_utils/os.py`. Also the line number seems to reference the state of `os.py` in the master branch.

Reply:
> ```
> TypeError: argument of type 'NoneType' is not iterable
> ```
> 
> And this is just the missing ocrd_tool that should come from the self._metadata from core v3.0.0

That's odd – 
```python
self._metadata = json.loads(resource_string(self.__module__.split('.')[0], 'ocrd-tool.json'))
```
– should work in ocrd_cis, too. Perhaps we are still missing ocrd-tool.json in MANIFEST.in?

Reply:
> OK, can you make sure it works if you do pip install -U shapely?

Yes, it does update the version of shapely.

> Not sure I get what you mean with "the make was failing" - is it working now or do you still get that error.

Sorry for the confusion, I meant `make install` in ocrd_cis. How do I install it with a v3.0.0a1 core dependency when that release is still not out yet?

Reply:
> How do I install it with a v3.0.0a1 core dependency when that release is still not out yet?

By pulling recent OCR-D/core#1240 (which does have the version) and installing it locally.

CI is of course not possible just yet. (In https://github.com/bertsky/core/pull/8 we will do some final changes before manually publishing today.)

Does `make test` work after adding the tool json to the manifest and reinstalling?

Reply:
I get this now:
```bash
(venv38-core) mm@MM-Notebook:~/repos/ocrd_cis$ bash tests/run_alignment_test.bash 
...
  File "/home/mm/repos/core-bertsky/core/build/__editable__.ocrd-3.0.0a1-py3-none-any/ocrd/processor/base.py", line 286, in process_workspace
    for input_file_tuple in self.zip_input_files(on_error='abort'):
TypeError: zip_input_files() got an unexpected keyword argument 'on_error'
```

Weird, because that keyword argument is not missing. I have #1240 installed in the same environment.

Reply:
Yes, that's absolutely crazy. You have the right version, no doubt. 

In tests/processor/test_processor.py we even verify this works – and the [recent CI](https://github.com/OCR-D/core/actions/runs/10402365756) succeeds.

I have no idea what is going on here. Is that even from a type checker, or the regular Python interpreter?

Reply:
Indeed, very strange. What is `python --version` and `pip freeze`, so I can try to reproduce?

Reply:
And can you try in a new venv, with just core and ocrd_cis installed?

Reply:
> And can you try in a new venv, with just core and ocrd_cis installed?

Yes, I have already done that. There was another other missing import - Levenshtein. 
Should I add `'python-Levenshtein>=0.25.1'` in the setup.py or was that library supposed to come from core?

After I have manually installed that, I am stuck again on the same error.

```bash
(venv38-core-v3) mm@MM-Notebook:~/repos/ocrd_cis$ python --version
Python 3.8.16
(venv38-core-v3) mm@MM-Notebook:~/repos/ocrd_cis$ pip freeze
...
Levenshtein==0.25.1
...
numpy==1.24.4
-e git+ssh://git@github.com/bertsky/core.git@228272b6a4ee94795e8266af4182eacae38e713c#egg=ocrd
ocrd-cis @ file:///home/mm/repos/ocrd_cis
...
```

Reply:
It is strange because all other tests pass normally although invoking the same method and are supposed to fail. o.0

Reply:
I also noticed missing Levenshtein and also broken `get_ocrd_tool`. I'll send a PR for that after finishing https://github.com/bertsky/core/pull/8. And I'll try to reproduce then.

Reply:
Just do the broken `get_ocrd_tool` fix in the PR. The missing Levenshtein was fixed by Robert. Just the align import was not replaced.

Reply:
Oops! Indeed, I forgot to update the non-ocropy processors in that regard.